src/algebra/star/algebra.lean

```diff
@@ -15,31 +16,56 @@ Introduces the notion of a star algebra over a star ring.
 
 universes u v w
 
+open opposite
+
 /--
 A star algebra `A` over a star ring `R` is an algebra which is a star ring,
 and the two star structures are compatible in the sense
 `star (r • a) = star r • star a`.
 -/
--- Note that we take `star_ring A` as a typeclass argument, rather than extending it,
+-- Note that we take `star_monoid A` as a typeclass argument, rather than extending it,
 -- to avoid having multiple definitions of the star operation.
-class star_algebra (R : Type u) (A : Type v)
-  [comm_semiring R] [star_ring R] [semiring A] [star_ring A] [algebra R A] :=
+class star_algebra (R : Type u) (A : Type v) [has_star R] [has_star A] [has_scalar R A] :=
 (star_smul : ∀ (r : R) (a : A), star (r • a) = (@has_star.star R _ r) • star a)
 
 @[simp] lemma star_smul {R : Type u} {A : Type v}
-  [comm_semiring R] [star_ring R] [semiring A] [star_ring A] [algebra R A] [star_algebra R A]
+  [has_star R] [has_star A] [has_scalar R A] [star_algebra R A]
   (r : R) (a : A) :
   star (r • a) = star r • star a :=
 star_algebra.star_smul r a
 
+attribute [instance] ring_hom_inv_pair.of_ring_equiv
+
+/-- If `A` is a left- and right- module over a commutative `R` with compatible actions,
+then `star` is a semilinear equivalence. -/
+@[simps]
+def star_linear_equiv {R : Type u} {A : Type v}
+  [comm_semiring R] [star_ring R] [semiring A] [star_ring A]
+  [module R A] [star_algebra R A] :
+    A ≃ₛₗ[((star_ring_aut : ring_aut R) : R →+* R)] A :=
+{ to_fun := star,
+  map_smul' := star_smul,
+  .. star_add_equiv }
+
+/-- If `A` is a left- and right- module over a noncommutative `R` with compatible actions,
+then `star` is a semilinear equivalence. -/
+@[simps]
+def star_linear_equiv_nc {R : Type u} {A : Type v}
+  [semiring R] [star_ring R] [semiring A] [star_ring A]
+  [module R A] [star_algebra R A]
+  [module Rᵒᵖ A] (h : ∀ (r : R) (a : A), r • a = op r • a) :
+    A ≃ₛₗ[(↑(star_ring_equiv : R ≃+* Rᵒᵖ) : R →+* Rᵒᵖ)] A :=
```

Comment:
Hmm, I would actually prefer we don't define this noncommutative version at all, until we have some use case requiring it (same with `of_ring_equiv` in `src/algebra/ring/comp_typeclasses`).  I don't like the fact that this is not involutive (because of the `(Rᵒᵖ)ᵒᵖ` issue).  I also don't like that it is of a different type from `refl` in the case where the star instance is trivial (i.e. `star_ring_of_comm`) -- note that, by contrast, `star_linear_equiv` is defeq to `refl` when we have the trivial star instance.

I think there is a way to solve this with yet more typeclasses (introduce a typeclass on monoids `[is_opposite R S]` like I was suggesting above).  But that's a refactor for another day.

Reply:
I think this is worth keeping if only to show that working in this generality is clumsy, and to remind ourselves what pieces might need refactors. I think `of_ring_equiv` is worth keeping too, but it shouldn't ever be used with `local attribute [instance]`

Reply:
Let's get a third opinion.  I don't think we should add things which the existing machinery doesn't yet support and for which we **also** have no intended use!  This goes double for `of_ring_equiv`, which could send an intermediate user in very much the wrong direction: if they are using it, something is wrong.

Reply:
I've committed the change to remove `[instance]` to give a more contained picture on what I think we disagree on. Perhaps we should get #9484 wrapped up before getting another pair of eyes on this.

Reply:
I agree with @hrmacbeth that adding code with the intention that it shouldn't be used is rather strange. For now I think my preferred solution would be to remove the code, but add a detailed TODO at the top of the file regarding the noncommutative case and why we haven't done it yet.

Reply:
Perhaps the easiest approach is just to split the PR, and add the contentious one later.

Reply:
I agree with the proposal to first merge #9484 and secondly merge the commutative part of this PR, before returning to the discussion of whether `of_ring_equiv` and `star_linear_equiv_nc` should exist.